python_modules/dagster-test/dagster_test/toys/user_computed_data_versions/__init__.py

```diff
@@ -0,0 +1,129 @@
+"""Models a case where Dagster is used purely for orchestration and not data transformation.
+
+In this scenario, there is an asset graph defined in some external system. An asset factory is used
+to build Dagster `AssetsDefinition` objects corresponding to the nodes in the external system's
+asset graph. The factory queries the external system for dependency and code version info.
+
+Dagster forwards materialization requests to the external system. We pass an asset key as a string
+together with serialized provenance information for the last materialization of the asset on record.
+The backend determines what to do with this information-- if it determines that the provenance of
+the last materialization is not stale (different from the current state of the asset), then it will
+simply use the existing stored value of the asset. Otherwise, it will recompute the asset and store
+the result.
+
+The materialization endpoint of the external system returns a data version string and a flag
+indicating whether a memoized value was used. This information is passed to the Dagster framework by
+returning a `Nothing` `Output`.
+"""
+import json
+import warnings
+from typing import cast
+
+from dagster import (
+    AssetKey,
+    AssetsDefinition,
+    DataProvenance,
+    DataVersion,
+    Definitions,
+    ExperimentalWarning,
+    In,
+    Nothing,
+    OpDefinition,
+    OpExecutionContext,
+    Out,
+    Output,
+    observable_source_asset,
+)
+from dagster._core.definitions.unresolved_asset_job_definition import define_asset_job
+
+from .external_system import ExternalSystem
+
+warnings.filterwarnings("ignore", category=ExperimentalWarning)
+
+EXTERNAL_SYSTEM = ExternalSystem()
+
+
+def external_asset(key: AssetKey):
+    """Factory to build an `AssetsDefinition` object to represent an asset in an external system.
+
+    The factory queries the external system for dependency and code version info. The op attached
+    to the `AssetsDefinition` forwards materialization requests to the external system, sending an
+    asset key and provenance info for the last recorded materialization of the asset. The external
+    system responds with a data version and a flag indicating whether a memoized value was used for
+    the materialization (i.e. whether recomputation actually occurred). This is passed to the
+    dagster framework by packaging it on a returned `Output` object. The `Output` returns `Nothing`
+    because Dagster is not handling passing data in between ops in this scenario.
+
+    Args:
+        key (AssetKey):
+            The asset key for the represented asset.
+
+    Returns (AssetsDefinition):
+        An `AssetsDefinition` object representing the asset in the external system.
+    """
+
+    def fn(context: OpExecutionContext, *args):
+        # Keep the temp storage directory used by the backend to the current instance.
+        EXTERNAL_SYSTEM.set_storage_directory(context.instance.storage_directory())
```

Comment:
Recommend just directly constructing ExternalSystem here rather than doing this setter